modules/building-memcached-operator-using-osdk.adoc

```diff
@@ -3,441 +3,523 @@
+	// +kubebuilder:validation:Minimum=0
+	// Size is the size of the memcached deployment
+	Size int32 `json:"size"`
+}
+
+// MemcachedStatus defines the observed state of Memcached
+type MemcachedStatus struct {
+	// Nodes are the names of the memcached pods
+	Nodes []string `json:"nodes"`
+}
 ----
 
-.... Set up RBAC and create the `memcached-operator` manifests:
+.. Add the `+kubebuilder:subresource:status` marker to add a `status` subresource to the CRD manifest:
 +
-[source,terminal]
+[source,go]
 ----
-$ oc create -f deploy/role.yaml
+// Memcached is the Schema for the memcacheds API
+// +kubebuilder:subresource:status <1>
+type Memcached struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MemcachedSpec   `json:"spec,omitempty"`
+	Status MemcachedStatus `json:"status,omitempty"`
+}
 ----
+<1> Add this line.
 +
-[source,terminal]
-----
-$ oc create -f deploy/role_binding.yaml
-----
+This enables the controller to update the CR status without changing the rest of the CR object.
+
+.. Update the generated code for the resource type:
 +
 [source,terminal]
 ----
-$ oc create -f deploy/service_account.yaml
+$ make generate
 ----
 +
-[source,terminal]
-----
-$ oc create -f deploy/operator.yaml
-----
+[TIP]
+====
+After you modify a `*_types.go` file, you must run the `make generate` command to update the generated code for that resource type.
+====
++
+The above Makefile target invokes the `controller-gen` utility to update the `api/v1/zz_generated.deepcopy.go` file. This ensures your API Go type definitions implement the `runtime.Object` interface that all `Kind` types must implement.
 
-.... Verify that the `memcached-operator` deploy is up and running:
+. Generate and update CRD manifests:
 +
 [source,terminal]
 ----
-$ oc get deployment
+$ make manifests
 ----
 +
-.Example output
-[source,terminal]
-----
-NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
-memcached-operator       1         1         1            1           1m
-----
+This Makefile target invokes the `controller-gen` utility to generate the CRD manifests in the `config/crd/bases/cache.example.com_memcacheds.yaml` file.
 
-... _Option B:_ Running locally outside the cluster.
+.. Optional: Add custom validation to your CRD.
 +
-This method is preferred during development cycle to deploy and test faster.
+OpenAPI v3.0 schemas are added to CRD manifests in the `spec.validation` block when the manifests are generated. This validation block allows Kubernetes to validate the properties in a Memcached CR when it is created or updated.
 +
-Run the Operator locally with the default Kubernetes configuration file present at `$HOME/.kube/config`:
+As an Operator author, you can use Kubebuilder markers (annotations) to configure custom validations for your API. These markers must always have a `+kubebuilder:validation` prefix. For example, adding an enum-type specification can be done by adding the following marker:
```

Comment:
Is there a way to avoid the parentheses for (annotations)? If not, that's okay, the general guidance is to avoid it possible.